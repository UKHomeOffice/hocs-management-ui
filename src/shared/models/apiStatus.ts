export default interface ApiStatus {
    display: string;
    timeoutPeriod: number;
    type: string;
}
