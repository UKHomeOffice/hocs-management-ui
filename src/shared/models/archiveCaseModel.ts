export default interface ArchiveCaseModel {
    caseReference: string;
    deleted: boolean;
}
