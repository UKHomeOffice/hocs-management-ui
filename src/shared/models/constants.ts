export const ADD_CHILD_TOPIC_ERROR_DESCRIPTION = 'Something went wrong while adding the child topic. Please try again.';
export const ADD_STANDARD_LINE_ERROR_DESCRIPTION = 'Something went wrong while adding the standard line. Please try again.';
export const ADD_TEMPLATE_ERROR_DESCRIPTION = 'Something went wrong while adding the template. Please try again.';
export const ADD_UNIT_ERROR_DESCRIPTION = 'Something went wrong while adding the unit. Please try again.';
export const ADD_USER_ERROR_DESCRIPTION = 'Something went wrong while adding the following users. Please try again.';
export const ADD_USER_ERROR_TITLE = 'There was an error adding the users';
export const DUPLICATE_CHILD_TOPIC_DESCRIPTION = 'A child topic with that name already exists';
export const DUPLICATE_UNIT_DESCRIPTION = 'A unit with those details already exists';
export const EMPTY_SUBMIT_ERROR_DESCRIPTION = 'Please select some users before submitting.';
export const EMPTY_SUBMIT_ERROR_TITLE = 'No users selected';
export const EMPTY_SUBMIT_TEAM_ERROR_DESCRIPTION = 'Please select a team before submitting.';
export const EMPTY_SUBMIT_TEAM_ERROR_TITLE = 'No team selected';
export const EMPTY_SUBMIT_DRAFT_QA_TEAM_ERROR_DESCRIPTION = 'Select a team for Initial Draft and QA response stages.';
export const EMPTY_SUBMIT_PRIVATE_MINISTER_TEAM_ERROR_DESCRIPTION = 'Select a team for Private Office and Minister sign off stages.';
export const EMPTY_SUBMIT_TOPIC_ERROR_DESCRIPTION = 'Select a topic before submitting.';
export const EMPTY_SUBMIT_TOPIC_ERROR_TITLE = 'No topic selected';
export const GENERAL_ERROR_TITLE = 'Something went wrong';
export const LOAD_PARENT_TOPICS_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_TOPICS_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_TOPIC_ERROR_DESCRIPTION = 'There was an error retrieving the parent topics.  Please try refreshing the page.';
export const LOAD_CASE_TYPES_DESCRIPTION = 'There was an error retrieving the case types.  Please try refreshing the page.';
export const LOAD_TEAM_ERROR_DESCRIPTION = 'There was an error retrieving the team.  Please try refreshing the page.';
export const LOAD_TEAM_MEMBERS_ERROR_DESCRIPTION = 'There was an error retrieving the teams members.  Please try refreshing the page.';
export const LOAD_TEAMS_ERROR_DESCRIPTION = 'There was an error retrieving the teams.  Please try refreshing the page.';
export const LOAD_USERS_ERROR_DESCRIPTION = 'There was an error retrieving the list of users.  Please try refreshing the page.';
export const REMOVE_FROM_TEAM_ALLOCATED_ERROR_DESCRIPTION = 'The user cannot be removed from the team as they have cases assigned';
export const REMOVE_FROM_TEAM_ERROR_DESCRIPTION = 'There was an error removing the user from the team';
export const VALIDATION_ERROR_TITLE = "Thereâ€™s a problem";
