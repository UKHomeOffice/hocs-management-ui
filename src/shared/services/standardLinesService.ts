import axios from 'axios';
import UpdateRequest from 'shared/pages/standardLineAmend/updateRequest';

export const addStandardLine = (standardLine: FormData) => new Promise((resolve, reject) => axios
    .post('/api/standard-lines', standardLine)
    .then(() => resolve())
    .catch(reason => reject(reason))
);

export const updateStandardLine = (standardLine: UpdateRequest) => new Promise((resolve, reject) => axios
    .post('/api/standard-lines/update', standardLine)
    .then(() => resolve())
    .catch(reason => reject(reason))
);

export const getAllStandardLines = () => new Promise((resolve, reject) => axios
    .get('/api/standard-lines/all')
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const getStandardLine = (standardLineUuid: string) => new Promise((resolve, reject) => axios
    .get(`/api/standard-lines/get/${standardLineUuid}`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);

export const expireStandardLine = (standardLineUuid: string) => new Promise((resolve, reject) => axios
    .post(`/api/standard-lines/expire/${standardLineUuid}`, {})
    .then(() => resolve())
    .catch(reason => reject(reason))
);

export const deleteStandardLine = (standardLineUuid: string) => new Promise((resolve, reject) => axios
    .delete(`/api/standard-lines/delete/${standardLineUuid}`, {})
    .then(() => resolve())
    .catch(reason => reject(reason))
);

export const downloadStandardLine = (standardLineUuid: string) => new Promise((resolve, reject) => axios
    .get(`/api/standard-lines/download/${standardLineUuid}`)
    .then(response => resolve(response.data))
    .catch(reason => reject(reason))
);
