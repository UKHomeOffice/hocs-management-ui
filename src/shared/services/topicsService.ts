import axios from 'axios';
import Item from '../models/item';

export const addChildTopic = (parentTopicId: string, displayName: string) => new Promise((resolve, reject) => axios
    .post(`/api/topics/parents/${parentTopicId}`, { displayName })
    .then(() => resolve())
    .catch(reason => reject(reason))
);

export const getParentTopics = () => new Promise((resolve, reject) => axios
    .get<Item[]>('/api/topics/parents')
    .then(response => resolve(response.data.sort((a, b) => {
        const labelA = a.label.toLocaleLowerCase();
        const labelB = b.label.toLocaleLowerCase();
        let comparison = 0;
        if (labelA > labelB) {
            comparison = 1;
        } else if (labelA < labelB) {
            comparison = -1;
        }
        return comparison;
    })))
    .catch(reason => reject(reason))
);
