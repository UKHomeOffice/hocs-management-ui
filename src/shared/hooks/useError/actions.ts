import { FormError } from '../../models/formError';

export type AddFormError = {
    payload: FormError;
    type: 'AddFormError';
};

export type ClearError = {
    type: 'ClearError'
};

export type SetTitle = {
    payload: string;
    type: 'SetTitle';
};

export type SetDescription = {
    payload: string;
    type: 'SetDescription';
};

export type Action =
    AddFormError |
    ClearError |
    SetDescription |
    SetTitle;
