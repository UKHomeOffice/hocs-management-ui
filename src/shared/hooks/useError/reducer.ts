import { Action } from './actions';
import PageError from '../../models/pageError';

export const reducer = (state: PageError, action: Action): PageError => {
    switch (action.type) {
        case 'AddFormError':
            return {
                error: { ...state.error, formErrors: [...state.error ? state.error.formErrors || [] : [], action.payload] }
            };
        case 'ClearError':
            return { error: undefined };
        case 'SetDescription':
            return { error: { ...state.error, description: action.payload } };
        case 'SetTitle':
            return { error: { ...state.error, title: action.payload } };
    }
    return state;
};
