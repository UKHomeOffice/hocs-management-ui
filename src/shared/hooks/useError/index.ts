import React, { Reducer } from 'react';
import PageError from '../../models/pageError';
import { Action } from './actions';
import { reducer } from './reducer';
import { FormError } from '../../models/formError';
import ErrorMessage from 'shared/models/errorMessage';

const useError = (formErrorDescription?: string, formErrorTitle?: string)
    : [
        PageError,
        (formError: FormError) => void,
        () => void,
        (error: ErrorMessage) => void
    ] => {
    const [pageError, dispatch] = React.useReducer<Reducer<PageError, Action>>(reducer, {});

    const addFormError = (formError: FormError) => {
        (formErrorDescription || formErrorTitle) && setMessage({ title: formErrorTitle || '', description: formErrorDescription || '' });
        dispatch({ type: 'AddFormError', payload: formError });
    };
    const clearErrors = () => dispatch({ type: 'ClearError' });
    const setMessage = (errorMessage: ErrorMessage) => {
        dispatch({ type: 'SetDescription', payload: errorMessage.description });
        dispatch({ type: 'SetTitle', payload: errorMessage.title });
    };
    return [pageError, addFormError, clearErrors, setMessage];
};

export default useError;
