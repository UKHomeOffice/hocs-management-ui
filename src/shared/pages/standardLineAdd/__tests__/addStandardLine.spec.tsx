import React from 'react';
import { match, MemoryRouter } from 'react-router-dom';
import { createBrowserHistory, History, Location } from 'history';
import { act, render, RenderResult, wait, fireEvent, waitForElement } from '@testing-library/react';
import AddStandardLine from '../addStandardLine';
import * as StandardLinesService from '../../../services/standardLinesService';
import { GENERAL_ERROR_TITLE, ADD_STANDARD_LINE_ERROR_DESCRIPTION, LOAD_TOPICS_ERROR_DESCRIPTION } from '../../../models/constants';
import StandardLine from '../../../models/standardLine';
import * as useError from '../../../hooks/useError';
import * as TopicsService from '../../../services/topicsService';
import { createMockFile } from '../../../../../test/createMockFile';
import { advanceTo } from 'jest-date-mock';

jest.mock('../../../services/topicsService', () => ({
    __esModule: true,
    getTopics: () => Promise.resolve([{
        label: '__topic1__',
        value: '__topicId1__'
    }, {
        label: '__topic2__',
        value: '__topicId2__'
    }])
}));

const getTopicsSpy = jest.spyOn(TopicsService, 'getTopics');

let match: match<any>;
let history: History<any>;
let location: Location;
let mockStandardLine: StandardLine;
let wrapper: RenderResult;

const useReducerSpy = jest.spyOn(React, 'useReducer');
const reducerDispatch = jest.fn();
const useErrorSpy = jest.spyOn(useError, 'default');
const addFormErrorSpy = jest.fn();
const clearErrorsSpy = jest.fn();
const setMessageSpy = jest.fn();

const renderComponent = () => render(
    <MemoryRouter>
        <AddStandardLine history={history} location={location} match={match}></AddStandardLine>
    </MemoryRouter>
);

jest.spyOn(StandardLinesService, 'addStandardLine').mockImplementation(() => Promise.resolve());

beforeEach(() => {
    history = createBrowserHistory();
    match = {
        isExact: true,
        params: { teamId: '__teamId__' },
        path: '',
        url: ''
    };

    location = {
        hash: '',
        key: '',
        pathname: '',
        search: '',
        state: {}
    };
    mockStandardLine = {
        expiryDate: '',
        topic: undefined,
        files: undefined
    };
    useReducerSpy.mockImplementation(() => [mockStandardLine, reducerDispatch]);
    useErrorSpy.mockImplementation(() => [{}, addFormErrorSpy, clearErrorsSpy, setMessageSpy]);
    history.push = jest.fn();
    reducerDispatch.mockReset();
    addFormErrorSpy.mockReset();
    clearErrorsSpy.mockReset();
    setMessageSpy.mockReset();
    advanceTo(new Date(1640995200000)); // reset to date time.
    act(() => {
        wrapper = renderComponent();
    });
});

describe('when the addStandardLine component is mounted', () => {
    it('should render with default props', async () => {
        expect.assertions(1);

        await wait(() => {
            expect(wrapper.container).toMatchSnapshot();
        });
    });

    it('should display an error if the call to retrieve the parent topics fails', async () => {
        expect.assertions(1);
        getTopicsSpy.mockImplementation(() => Promise.reject('error'));

        act(() => {
            wrapper = renderComponent();
        });

        await wait(() => {
            expect(setMessageSpy).toBeCalledWith({ title: GENERAL_ERROR_TITLE, description: LOAD_TOPICS_ERROR_DESCRIPTION });
        });

    });
});

describe('when the submit button is clicked', () => {
    describe('and the data is filled in', () => {

        beforeEach(async () => {
            mockStandardLine.expiryDate = '2101-01-01';
            mockStandardLine.files = [createMockFile()];
            mockStandardLine.topic = { label: '__topic__', value: '__value__' };
            const submitButton = await waitForElement(async () => {
                return await wrapper.findByText('Submit');
            });

            fireEvent.click(submitButton);
        });

        describe('and the service call is successful', () => {
            it('should redirect to the home page', async () => {
                expect.assertions(1);

                await wait(() => {
                    expect(history.push).toHaveBeenCalledWith('/manage-standard-lines', { successMessage: 'The standard line was created successfully' });
                });
            });
            it('should clear any previous errors', async () => {
                expect.assertions(1);

                await wait(() => {
                    expect(clearErrorsSpy).toHaveBeenCalled();
                });
            });
        });

        describe('and the service call fails', () => {
            beforeAll(() => {
                jest.spyOn(StandardLinesService, 'addStandardLine').mockImplementationOnce(() => Promise.reject({ response: { status: 500 } }));
            });

            it('should set the error state', () => {
                expect(setMessageSpy).toHaveBeenCalledWith({ description: ADD_STANDARD_LINE_ERROR_DESCRIPTION, title: GENERAL_ERROR_TITLE });
            });

            it('should clear any previous errors', () => {
                expect(clearErrorsSpy).toHaveBeenCalled();
            });
        });
    });
    describe('and the data is not filled in', () => {
        beforeEach(async () => {
            const submitButton = await waitForElement(async () => {
                return await wrapper.findByText('Submit');
            });

            fireEvent.click(submitButton);
        });

        it('should call the begin submit action', () => {
            expect(clearErrorsSpy).toHaveBeenCalled();
        });

        it('should set the error state', () => {
            expect(addFormErrorSpy).toHaveBeenNthCalledWith(1, { key: 'files', value: 'The Standard Line is required' });
            expect(addFormErrorSpy).toHaveBeenNthCalledWith(2, { key: 'expiryDate', value: 'The Expiry Date is invalid' });
            expect(addFormErrorSpy).toHaveBeenNthCalledWith(3, { key: 'topic', value: 'The Topic is required' });
        });
    });
    describe('and the expiry date is not in the future', () => {
        beforeEach(async () => {
            mockStandardLine.expiryDate = new Date().toDateString();
            mockStandardLine.files = [createMockFile()];
            mockStandardLine.topic = { label: '__topic__', value: '__value__' };
            const submitButton = await waitForElement(async () => {
                return await wrapper.findByText('Submit');
            });
            fireEvent.click(submitButton);
        });

        it('should show the validation message', () => {
            expect(addFormErrorSpy).toHaveBeenNthCalledWith(1, { key: 'expiryDate', value: 'The Expiry Date must be in the future' });
        });
    });
});
