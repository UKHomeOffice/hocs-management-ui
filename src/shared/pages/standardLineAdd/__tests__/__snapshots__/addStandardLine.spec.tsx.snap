// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when the addStandardLine component is mounted should render with default props 1`] = `
<div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-two-thirds-from-desktop"
    >
      <a
        class="govuk-back-link"
        href="/"
      >
        Back
      </a>
      <h1
        class="govuk-heading-xl"
      >
        Add a Standard Line
      </h1>
    </div>
  </div>
  <div
    class="govuk-grid-row"
  >
    <div
      class="govuk-grid-column-one-half-from-desktop"
    >
      <form
        action="/api/standardLines"
        method="POST"
      >
        <input
          name="_csrf"
          type="hidden"
          value=""
        />
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label govuk-label--s"
            for="topics-input"
            id="topics-label"
          >
            Topics
          </label>
          <div
            class=" css-b62m3t-container"
            id="topics"
          >
            <span
              class="css-1f43avz-a11yText-A11yText"
              id="react-select-2-live-region"
            />
            <span
              aria-atomic="false"
              aria-live="polite"
              aria-relevant="additions text"
              class="css-1f43avz-a11yText-A11yText"
              role="log"
            />
            <div
              class="govuk-typeahead__control css-1g8ily9-Control"
            >
              <div
                class="govuk-typeahead__value-container css-pilx6v-ValueContainer"
              >
                <div
                  class="govuk-typeahead__placeholder css-ovmm9z-Placeholder"
                  id="react-select-2-placeholder"
                >
                  Search
                </div>
                <div
                  class="govuk-typeahead__input-container css-qbdosj-Input"
                  data-value=""
                >
                  <input
                    aria-activedescendant=""
                    aria-autocomplete="list"
                    aria-describedby="react-select-2-placeholder"
                    aria-expanded="false"
                    aria-haspopup="true"
                    autocapitalize="none"
                    autocomplete="off"
                    autocorrect="off"
                    class="govuk-typeahead__input"
                    id="topics-input"
                    role="combobox"
                    spellcheck="false"
                    style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                    tabindex="0"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class="govuk-typeahead__indicators css-1hb7zxy-IndicatorsContainer"
              >
                <div
                  aria-hidden="true"
                  class="govuk-typeahead__indicator govuk-typeahead__loading-indicator css-pogzpp-loadingIndicator"
                >
                  <span
                    class="css-1xtdfmb-LoadingDot"
                  />
                  <span
                    class="css-zoievk-LoadingDot"
                  />
                  <span
                    class="css-x748d8-LoadingDot"
                  />
                </div>
                <span
                  class="govuk-typeahead__indicator-separator css-62wm9a-IndicatorSeparator"
                />
                <div
                  aria-hidden="true"
                  class="govuk-typeahead__indicator govuk-typeahead__dropdown-indicator css-hbt8bd-DropdownIndicator"
                >
                  <svg
                    aria-hidden="true"
                    class="css-tj5bde-Svg"
                    focusable="false"
                    height="20"
                    viewBox="0 0 20 20"
                    width="20"
                  >
                    <path
                      d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                    />
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="govuk-form-group"
        >
          <label
            class="govuk-label"
            for="files"
            id="files-label"
          >
            <label
              class="govuk-label govuk-label--s"
              for="files"
              id="files-label"
            >
              Add document
            </label>
          </label>
          <input
            class="govuk-file-upload"
            id="files"
            name="files"
            type="file"
          />
        </div>
        <div
          class="govuk-form-group"
        >
          <fieldset
            class="govuk-fieldset"
            id="expiryDate"
            role="group"
          >
            <legend
              class="govuk-fieldset__legend govuk-label--s"
            >
              Expiry Date
            </legend>
            <div
              class="govuk-date-input"
            >
              <div
                class="govuk-date-input__item"
              >
                <div
                  class="govuk-form-group"
                >
                  <label
                    class="govuk-label govuk-date-input__label"
                    for="expiryDate-day"
                  >
                    Day
                  </label>
                  <input
                    class="govuk-input govuk-date-input__input govuk-input--width-2 "
                    id="expiryDate-day"
                    inputmode="numeric"
                    max="31"
                    min="1"
                    name="expiryDate-day"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <div
                class="govuk-date-input__item"
              >
                <div
                  class="govuk-form-group"
                >
                  <label
                    class="govuk-label govuk-date-input__label"
                    for="expiryDate-month"
                  >
                    Month
                  </label>
                  <input
                    class="govuk-input govuk-date-input__input govuk-input--width-2 "
                    id="expiryDate-month"
                    max="12"
                    min="1"
                    name="expiryDate-month"
                    pattern="[0-9]*"
                    type="number"
                    value=""
                  />
                </div>
              </div>
              <div
                class="govuk-date-input__item"
              >
                <div
                  class="govuk-form-group"
                >
                  <label
                    class="govuk-label govuk-date-input__label"
                    for="expiryDate-year"
                  >
                    Year
                  </label>
                  <input
                    class="govuk-input govuk-date-input__input govuk-input--width-4  "
                    id="expiryDate-year"
                    max="2122"
                    min="1900"
                    name="expiryDate-year"
                    pattern="[0-9]*"
                    type="number"
                    value=""
                  />
                </div>
              </div>
            </div>
          </fieldset>
        </div>
        <input
          class="govuk-button"
          type="submit"
          value="Submit"
        />
      </form>
    </div>
  </div>
</div>
`;
